### DISCUSSION TITLE
Kotlin 2.0 (Kotlin Symbol Processing) support

### DISCUSSION THREAD

**Discussion by milgner** [pf4j/pf4j#547] [Ideas]
The documentation currently mentions using kapt for annotation processing in Kotlin. In Kotlin 2.0, namely the new K2 compiler version, kapt will be replaced with a non-backwards-compatible API called Kotlin Symbol Processing (KSP, not to be confused with Kerbal Space Program :wink:).

I'd like to write a KSP-based implementation for processing PF4J annotations. Would you prefer integrating them into the main repository (using a regular pull request) or should I create a separate repository / module for that?

**Comment by decebals** [comment]
I propose you to create a separate repository for the moment. The reason is I don't use Kotlin and for me it's hard (time consuming) to maintain a such thing.

**Comment by delopst** [comment]
Hi.

I did not see the discussion here. Sorry for that. I had the same idea and since we were in a hurry (there were no existing implementations yet), I already published my implementation here: [pf4j-kotlin-symbol-processing](https://github.com/better-care/pf4j-kotlin-symbol-processing).

I am open to any discussion about transferring the repository to the pf4j account if needed. @milgner I can also offer some maintenance of the plugin to be compatible with the latest Kotlin versions since I work in Kotlin a lot.

For duplication reasons, I am closing the issue but I am linking it here if needed: [#522](https://github.com/pf4j/pf4j/issues/552)

Regards, Primož

**Reply by milgner** [reply]
That is great! I see that it's currently implemented as a Maven plugin so I'll want to add some glue code for a Gradle plugin, too.
But really great to know that there is shared interest!
I don't know when exactly I'll be able to get around to checking it out but wanted to express my thanks already 😊

**Reply by delopst** [reply]
We were in a hurry and I only tested this with maven. I have a plan to test this in grade and add some examples by the end of this week. If you will have some time before that, please let me know :) 

We do not want to duplicate the work :) 

Thank you :)

**Reply by delopst** [reply]
I have tested the build with Gradle and should work out of the box. Extensions.idx file was generated and it was correct.
See the following example:

Define Gradle plugin portal to provide easy access to both core and community plugins:

```
pluginManagement {
    repositories {
        gradlePluginPortal()
    }
}
```

Apply the com.google.devtools.ksp plugin with the specified version and pf4j-kotlin-symbol-processor to the list of dependencies:

```
plugins {
    id("com.google.devtools.ksp") version "1.9.10-1.0.13"
}

dependencies {
    implementation("org.pf4j:pf4j:3.3.1")
    ksp("care.better.pf4j:pf4j-kotlin-symbol-processing:1.0.0-RC1")
}
```

I have also added this to README.md
